[MERGE] hr_attendance: Compute overtime on attendances

Williambraecky opened the pull request:
Adds a new feature to count overtime when using the attendance app.

Also adds an app that links the attendance and the holidays app to allow users to convert overtime hours to time off time.

Continuation of odoo/odoo#65542

Task ID: 1904850

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/69930).

nseinlet commented on addons/hr_attendance/report/hr_attendance_report.py:
```diff
@@ -0,0 +1,51 @@
+    overtime_hours = fields.Float("Extra Hours", readonly=True)
+
+    def init(self):
+        tools.drop_view_if_exists(self.env.cr, self._table)
+
+        self.env.cr.execute("""
+            CREATE OR REPLACE VIEW %s AS (
```
isn't combination of both a bit overkill?

Williambraecky replied:
I agree it's redundant, though it is present in a lot of report in the codebase, I will remove it.

jdoutreloux commented on addons/hr_holidays_attendance/models/hr_leave.py:
```diff
@@ -0,0 +1,89 @@
+            employee_sudo = self.env['hr.employee'].sudo().browse(vals['employee_id']) if 'employee_id' in vals else None
+            for leave in self.filtered('overtime_id'):
+                employee = employee_sudo if employee_sudo else leave.employee_id
+                calendar = employee.resource_calendar_id or self.env.company.resource_calendar_id
```
To avoid multi-company issues, you shouldn't use self.env.company

Williambraecky replied:
Changed to:

`                calendar = employee.resource_calendar_id or employee.company_id.resource_calendar_id`

jdoutreloux commented on addons/hr_holidays_attendance/models/hr_leave_allocation.py:
```diff
@@ -0,0 +1,83 @@
+        return res
+
+    @api.model
+    def create(self, vals):
```
Could you try to make this work in batch please?

Williambraecky replied:
Parent model also have api.model on create, I will check if we can replace it with a multicreate

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,183 @@ def _check_validity(self):
                         'datetime': format_datetime(self.env, last_attendance_before_check_out.check_in, dt_format=False),
                     })
 
+    def _get_attendances_dates(self):
+        # Returns a dictionnary {employee_id: set(dates)}
+        attendances_emp = defaultdict(set)
+        for attendance in self.filtered(lambda a: a.employee_id.company_id.hr_attendance_overtime and a.check_in):
+            tz = pytz.timezone(attendance.employee_id.tz or 'UTC')
+            check_in = pytz.utc.localize(attendance.check_in).astimezone(tz).replace(tzinfo=None)
+            if check_in < datetime.combine(attendance.employee_id.company_id.overtime_start_date, datetime.min.time()):
+                continue
+            attendances_emp[attendance.employee_id].add(check_in.date())
+            if attendance.check_out:
+                check_out = pytz.utc.localize(attendance.check_out).astimezone(tz).replace(tzinfo=None)
+                attendances_emp[attendance.employee_id].add(check_out.date())
+        return attendances_emp
```
```suggestion
def _get_day_start(self, employee, dt):
    date_employee_tz = pytz.utc.localize(dt).astimezone(employee._get_tz())
    start_day_employee_tz = date_employee_tz.replace(hour=0, minute=0, second=0)
    return start_day_employee_tz.astimezone(pytz.utc).replace(tzinfo=None)
    
def _get_attendances_dates(self):
        # Returns a dictionnary {employee_id: set(dates)}
        attendances_emp = defaultdict(set)
        for attendance in self.filtered(lambda a: a.employee_id.company_id.hr_attendance_overtime and a.check_in):
            check_in = attendance._get_day_start(attendance.employee_id, attendance.check_in)
            if check_in <= datetime.combine(attendance.employee_id.company_id.overtime_start_date, datetime.min.time()):
                continue
            attendances_emp[attendance.employee_id].add(check_in.date())
            if attendance.check_out:
                check_out = attendance._get_day_start(attendance.employee_id, attendance.check_out)
                attendances_emp[attendance.employee_id].add(check_out.date())
        return attendances_emp
```

and in hr.employee, something like :
```python
def _get_tz(self):
        return self.tz or\
            self.resource_calendar_id.tz or\
            self.company_id.resource_calendar_id.tz or\
            'UTC'
```

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,183 @@ def _check_validity(self):
+            tz = pytz.timezone(emp.tz or 'UTC')
+            attendance_domain = []
+            for attendance_date in attendance_dates:
+                start = datetime.combine(attendance_date, datetime.min.time())
+                # Translate back to dates stored in database
+                start_loc = tz.localize(start).astimezone(pytz.utc).replace(tzinfo=None)
+                attendance_domain = OR([attendance_domain, [
+                    ('check_in', '>=', start_loc), ('check_in', '<', start_loc + timedelta(hours=24)),
+                ]])
```
```suggestion
                attendance_domain = OR([attendance_domain, [
                    ('check_in', '>=', attendance_date), ('check_in', '<', attendance_date + timedelta(hours=24)),
                ]])
```

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,183 @@ def _check_validity(self):
+            attendances_per_day = defaultdict(lambda: self.env['hr.attendance'])
+            all_attendances = self.env['hr.attendance'].search(attendance_domain)
+            for attendance in all_attendances:
+                check_in = pytz.utc.localize(attendance.check_in).astimezone(tz).replace(tzinfo=None)
+                attendances_per_day[check_in.date()] += attendance
```
```suggestion
                check_in = attendance._get_day_start(attendance.employee_id, attendance.check_in)
                attendances_per_day[check_in] += attendance
```

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,183 @@ def _check_validity(self):
+                check_in = pytz.utc.localize(attendance.check_in).astimezone(tz).replace(tzinfo=None)
+                attendances_per_day[check_in.date()] += attendance
+
+            start = tz.localize(datetime.combine(min(attendance_dates), datetime.min.time())).astimezone(pytz.utc)
+            stop = tz.localize(datetime.combine(max(attendance_dates), datetime.max.time())).astimezone(pytz.utc)
```
```suggestion
            start = min(attendance_dates)
            stop = max(attendance_dates) + timedelta(hours=24)
```

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,183 @@ def _check_validity(self):
+                # Exclude resource.calendar.attendance
+                working_times[expected_attendance[0].date()].append(expected_attendance[:2])
+
+            overtimes = self.env['hr.attendance.overtime'].sudo().search([
+                ('employee_id', '=', emp.id),
+                ('date', 'in', list(attendance_dates)),
+                ('adjustment', '=', False),
+            ])
```
```suggestion
            attendance_dates_list = [attendance_date.astimezone(emp._get_tz()).date() for attendance_date in attendance_dates]
            overtimes = self.env['hr.attendance.overtime'].sudo().search([
                ('employee_id', '=', emp.id),
                ('date', 'in', attendance_dates),
                ('adjustment', '=', False),
            ])
```

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,183 @@ def _check_validity(self):
+                    if not overtime and overtime_duration:
+                        overtime_vals_list.append({
+                            'employee_id': emp.id,
+                            'date': attendance_date,
```
```suggestion
'date': attendance.astimezone(emp._get_tz()).date(),
```

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,183 @@ def _check_validity(self):
+            working_times = defaultdict(lambda: [])
+            for expected_attendance in expected_attendances:
+                # Exclude resource.calendar.attendance
+                working_times[expected_attendance[0].date()].append(expected_attendance[:2])
```
```suggestion
self._get_day_start(emp, expected_attendance[0])
```

tlibioulle reviewed (commented):
Hi @Williambraecky 

As discussed, here is a small suggestion. Do what you think is best with it :) (For sure you may trash it :) )

cc @tivisse : What do you think of it ?
The goal is mainly to add a util method which returns a naive datetime representing the start of the day of the employee, and deal with it along all the _update_overtime method.

Williambraecky commented:
@tivisse I fixed the issue with the failing test, do you want me to rebase those messy commits?

> Hi @Williambraecky
> 
> As discussed, here is a small suggestion. Do what you think is best with it :) (For sure you may trash it :) )
> 
> cc @tivisse : What do you think of it ?
> The goal is mainly to add a util method which returns a naive datetime representing the start of the day of the employee, and deal with it along all the _update_overtime method.

Hello @tlibioulle, I haven't applied all of your suggestion yet, however I added a timezone 'delta' cache to avoid localizing and converting timezones all the time.

I guess the only issue with that system would be summer/winter time changes however I do believe this is a non issue as most people don't work around midnight + time changes usually don't change the day (i e 1h becomes 0h or 1h becomes 2h). What do you think?

nseinlet commented on addons/hr_attendance/models/hr_attendance_overtime.py:
```diff
@@ -0,0 +1,32 @@
+    adjustment = fields.Boolean(default=False)
+
+    def init(self):
+        # Allows only 1 overtime record per employee per day unless it's an adjustment
+        self.env.cr.execute("""
+            CREATE UNIQUE INDEX IF NOT EXISTS hr_attendance_overtime_unique_employee_per_day
+            ON %s (employee_id, date)
+            WHERE adjustment is false""" % (self._table))
```
tools.create_unique_index

Williambraecky replied:
@nseinlet Hello, we looked into it with Yannick, IIRC `create_unique_index` doesn't allow for condition on the index, which we want here

tlibioulle commented on addons/hr_attendance/models/hr_attendance.py:
```diff
@@ -99,6 +104,194 @@ def _check_validity(self):
+        attendances_emp = defaultdict(set)
+        for attendance in self.filtered(lambda a: a.employee_id.company_id.hr_attendance_overtime and a.check_in):
+            tz = attendance.employee_id.tz or 'UTC'
+            check_in = attendance.check_in + tz_delta_cache[tz]
```
So we'll deal with naïve datetime in a different timezone than UTC, right. It looks really weird to me.

Williambraecky replied:
The handling of dates is essentially the same as before, we just avoid rewriting the whole `localize`, `astimezone` and `replace(tzinfo..)` every time, if we could query using timezones directly this would be easily avoided but unfortunately that's not the case ..

As for `Second, we have many calls to datetime.date(), which means we delegate the responsibility to the python library to return the date of a naive datetime.`
I am not quite sure what the issue would be with the date() function calls, could you elaborate? At first glance it seems like a really simple function here: https://github.com/python/cpython/blob/main/Lib/datetime.py#L1810-L1812

I will take another fresh look at all of this I am sure we can find a solution that will please both of us 😄

tlibioulle replied:
The problem I noticed was (related to the fact we deal with naive non-UTC datetime) : You want the day of a naive datetime, so we know which day we are, but we don't know where on earth (we have to read the whole code to understand the "where")?

So I think what you proposed at last is a solution more clear and understandable, we don't have to read the whole code to understand what we are doing with the datetime/date/timezones. (& it makes me happy :) )

tlibioulle reviewed (commented):
After a second check to the updated solution, my POV is that we may face some difficulties while maintaining the attendance computation part of code, talking about dates and timezones.

First, we deal with naive datetimes which are not in the UTC tz, which is not straightforward. I would either use aware datetimes or naive datetime which are unofficially in utc.
Second, we have many calls to datetime.date(), which means we delegate the responsibility to the python library to return the date of a naive datetime.

I'm not really convinced this is a better pattern than the previous one. I would have tried something with less back&forth between datetime and dates, and dealing with naive datetimes with the notion of timezone (when we add utcoffset manually rather than relying on aware datetimes) is something that causes me nightmares.

So, I keep the same mind than my previous comments : I would rather use one method to have the datetime (UTC, naive) representing the start of the day given the tz of the employee. Deal with it along the way and transform it in date only twice : when querying attendances and creating attendances.

Nevertheless I do think this is an opinion and if you're not convinced by my opinion and do think this is not an improvement to the design of your solution, it's not a big deal to me :)

Williambraecky commented:
@tlibioulle Hey, can you take a look at the latest version? 😄 
I believe this is the best of both versions and actually avoids even more conversion 🥳

tlibioulle commented:
> @tlibioulle Hey, can you take a look at the latest version? 
> I believe this is the best of both versions and actually avoids even more conversion

@Williambraecky You convinced me, great solution imho !

tivisse commented:
@robodoo r+ merge

robodoo commented:
Merge method set to merge directly, using the PR as merge commit message

robodoo commented:
Linked pull request(s) odoo/upgrade#2429 not ready. Linked PRs are not staged until all of them are ready.
